bundles/org.openhab.binding.freeathomesystem/src/main/java/org/openhab/binding/freeathomesystem/internal/FreeAtHomeSystemDiscoveryService.java

```diff
@@ -0,0 +1,134 @@
+ * The {@link FreeAtHomeSystemDiscoveryService} is responsible for performing discovery of things
+ *
+ * @author Andras Uhrin - Initial contribution
+ */
+@NonNullByDefault
+public class FreeAtHomeSystemDiscoveryService extends AbstractDiscoveryService implements ThingHandlerService {
+
+    private final Logger logger = LoggerFactory.getLogger(FreeAtHomeSystemDiscoveryService.class);
+
+    private @NonNullByDefault({}) FreeAtHomeBridgeHandler bridgeHandler;
+
+    Runnable runnable = new Runnable() {
+        @Override
+        public void run() {
+            if (bridgeHandler != null) {
+                ThingUID bridgeUID = bridgeHandler.getThing().getUID();
+
+                List<String> deviceList;
+                try {
+                    deviceList = bridgeHandler.getDeviceDeviceList();
+
+                    for (int i = 0; i < deviceList.size(); i++) {
+                        FreeAtHomeDeviceDescription device = bridgeHandler
+                                .getFreeatHomeDeviceDescription(deviceList.get(i));
+
+                        boolean useGenericDevice = true;
+
+                        if (useGenericDevice) {
+                            ThingUID uid = new ThingUID(FreeAtHomeSystemBindingConstants.FREEATHOMEDEVICE_TYPE_UID,
+                                    bridgeUID, device.deviceId);
+                            Map<String, Object> properties = new HashMap<>(1);
+                            properties.put("deviceId", device.deviceId);
+                            properties.put("interface", device.interfaceType);
+
+                            String deviceLabel = device.deviceLabel;
+
+                            DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(uid).withLabel(deviceLabel)
+                                    .withBridge(bridgeUID).withProperties(properties).build();
+
+                            thingDiscovered(discoveryResult);
+
+                            logger.debug("Thing discovered - DeviceId: {} - Device label: {}", device.getDeviceId(),
+                                    device.getDeviceLabel());
+                        }
+                    }
+
+                    stopScan();
+                } catch (FreeAtHomeHttpCommunicationException e) {
+                    logger.debug("TCommunication error in device discovery with the bridge: {}",
+                            bridgeHandler.getThing().getLabel());
+                }
+            }
+        }
+    };
+
+    public FreeAtHomeSystemDiscoveryService(int timeout) {
+        super(FreeAtHomeSystemBindingConstants.SUPPORTED_THING_TYPES_UIDS, timeout, false);
+    }
+
+    public FreeAtHomeSystemDiscoveryService() {
+        super(FreeAtHomeSystemBindingConstants.SUPPORTED_THING_TYPES_UIDS, 90, false);
+    }
+
+    @Override
+    public void setThingHandler(ThingHandler handler) {
+        if (handler instanceof FreeAtHomeBridgeHandler bridgeHandler) {
+            this.bridgeHandler = bridgeHandler;
+        }
+    }
```

Comment:
With the latest refactoring, I don't think you need this anymore, since you should be able to access the bridge through `this.thingHandler`. See for example #16250.